elektra-web remote instance management

omnidan opened the pull request:
an API and web user interface to remotely manage multiple elektra instances

documentation / usage is provided in the corresponding README.md's

@markus2330 please review my pull request

**EDIT:** I refactored most of the code and added more comments, cluster and instance management are implemented now, the client is now served by elektrad and I added a quickstart guide to the main README.md (I haven't had a chance to look into the new comments yet)

markus2330 commented on src/tools/web/README.md:
```diff
@@ -0,0 +1,37 @@
+# elektra web
+
+_an API and web user interface to remotely manage multiple elektra instances_
```
Elektra capital E

markus2330 commented on src/tools/web/README.md:
```diff
@@ -0,0 +1,37 @@
+Elektra web consists of multiple components:
+
+ * (multiple) servers running an elektra daemon ([`elektrad`](elektrad/))
+ * a single cluster management server to communicate with the elektra daemons ([`clusterd`](clusterd/))
```
recursive nesting missing

omnidan replied:
I'm not sure if recursive nesting is such a common usecase and I want to focus on getting one-level clusters working properly (main use case), so I'll leave this out for now. If I were to implement it, I'd do it with a single `clusterd` instance, though (allowing clusters to have clusters as their children).

markus2330 replied:
We should discuss this in the use case PR.

markus2330 commented on src/tools/web/README.md:
```diff
@@ -0,0 +1,37 @@
+
+ * (multiple) servers running an elektra daemon ([`elektrad`](elektrad/))
+ * a single cluster management server to communicate with the elektra daemons ([`clusterd`](clusterd/))
+ * a client (web browser) that accesses the Web UI on the cluster management server ([`client`](client/))
```
or elektrad

omnidan replied:
the client actually doesn't access elektrad directly, clusterd acts as a kind of router that maintains instances (this makes it easy to have a client that can be accessed from various devices, otherwise I'd have to store the connection info for clusterd/elektrad instances in localstorage, cookies or something similar). this way all you have to provide in the client is the host of the clusterd instance.

markus2330 commented on src/tools/web/README.md:
```diff
@@ -0,0 +1,37 @@
+ * a single cluster management server to communicate with the elektra daemons ([`clusterd`](clusterd/))
+ * a client (web browser) that accesses the Web UI on the cluster management server ([`client`](client/))
+
+![https://cdn.rawgit.com/omnidan/libelektra/http-api-proposal/src/tools/web/doc/network_structure.png](https://cdn.rawgit.com/omnidan/libelektra/http-api-proposal/src/tools/web/doc/network_structure.png)
```
img missing

omnidan replied:
I changed the branch and forgot to update the links - thanks for the notice!

markus2330 commented on src/tools/web/README.md:
```diff
@@ -0,0 +1,37 @@
+single instances can be configured independently.
+
+The configuration view of elektra web is similar to the tree view of the
+[qt-gui](https://github.com/ElektraInitiative/libelektra/tree/master/src/tools/qt-gui).
```
prefer to link against `git.libelektra.org/..`

markus2330 commented on src/tools/web/README.md:
```diff
@@ -0,0 +1,37 @@
+The configuration view of elektra web is similar to the tree view of the
+[qt-gui](https://github.com/ElektraInitiative/libelektra/tree/master/src/tools/qt-gui).
+
+![https://cdn.rawgit.com/omnidan/libelektra/http-api-proposal/src/tools/web/doc/ui_structure.png](https://cdn.rawgit.com/omnidan/libelektra/http-api-proposal/src/tools/web/doc/ui_structure.png)
```
Not found, please stay within Elektra repo.

markus2330 commented on src/tools/web/client/README.md:
````diff
@@ -0,0 +1,17 @@
+## Installation
+
+```
+npm install
````
In which directory?

Tell where the dependencies are listed.

Installation fails with:
```
npm WARN package.json elektra-web-client@1.0.0 No description
npm WARN package.json elektra-web-client@1.0.0 No repository field.
npm WARN package.json webpack-dev-server@1.16.1 No README data
npm ERR! peerinvalid The package react does not satisfy its siblings' peerDependencies requirements!
npm ERR! peerinvalid Peer material-ui@0.15.4 wants react@^15.0.0
npm ERR! peerinvalid Peer react-dom@15.3.2 wants react@^15.3.2
npm ERR! peerinvalid Peer react-redux@4.4.5 wants react@^0.14.0 || ^15.0.0-0
npm ERR! peerinvalid Peer react-tap-event-plugin@1.0.0 wants react@^15.0.0-0
npm ERR! peerinvalid Peer react-treeview@0.4.5 wants react@>=0.14.0
npm ERR! peerinvalid Peer rgx@0.2.4 wants react@^0.14.0

npm ERR! System Linux 3.16.0-4-amd64
npm ERR! command "/usr/bin/nodejs" "/usr/bin/npm" "install"
npm ERR! cwd /home/markus/Projekte/Elektra/libelektra/src/tools/web/client
npm ERR! node -v v0.10.29
npm ERR! npm -v 1.4.21
npm ERR! code EPEERINVALID
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/markus/Projekte/Elektra/libelektra/src/tools/web/client/npm-debug.log
npm ERR! not ok code 0
```

markus2330 replied:
cmake integration?

omnidan replied:
> In which directory?

all commands should be run in the same directory as the README

> Tell where the dependencies are listed.

In `package.json` - this is a nodejs standard

> Installation fails with:

I just tried reinstalling and it worked fine for me. It might be your node version, which is very old (current release is 6.x)

> cmake integration?

should be possible, it just needs to execute some bash commands

markus2330 commented on src/tools/web/client/README.md:
````diff
@@ -0,0 +1,17 @@
+## Running
+
+```
+npm start
````
Is it possible to start without installation?

markus2330 commented on src/tools/web/client/css/react-treeview.css:
```diff
@@ -0,0 +1,42 @@
+  overflow-y: hidden;
+}
+
+.tree-view_item {
```
`-` and _ mixture?

omnidan replied:
I copied this from [`react-treeview`](https://github.com/chenglou/react-treeview) (which I based my treeview off), but I'll change it to use `-` only.

markus2330 commented on src/tools/web/client/css/react-treeview.css:
```diff
@@ -0,0 +1,42 @@
+  cursor: pointer;
+}
+
+.tree-view_arrow {
```
please comment more around here.

markus2330 commented on src/tools/web/client/index.html:
```diff
@@ -0,0 +1,14 @@
+  <head>
+    <meta charset="UTF-8">
+    <title>elektra-web</title>
+    <link href="https://fonts.googleapis.com/css?family=Roboto:300,400,500" rel="stylesheet">
```
avoid all external links, the webui is intended for internal use.

markus2330 commented on src/tools/web/client/src/components/CreateInstanceCard.jsx:
```diff
@@ -0,0 +1,73 @@
+                  hintText="my webserver"
+                  onChange={(evt) => this.setState({ name: evt.target.value })}
+                  value={this.state.name}
+                />
```
Please add more comments.

markus2330 replied:
Issue still open

markus2330 commented on src/tools/web/clusterd/package.json:
```diff
@@ -0,0 +1,45 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/omnidan/elektra-web.git"
```
Avoid links outside of Elektra repo please.

markus2330 commented on src/tools/web/clusterd/src/db.js:
```diff
@@ -0,0 +1,44 @@
+import low from 'lowdb'
```
avoid extra databases next to Elektra

omnidan replied:
I thought it would be easier to use a simple JSON based db here, but it turns out it's not. I would have to wrap the elektra configuration into a cluster configuration (in lowdb) and unwrap it again later, and then special cases aren't handled yet (like metadata, etc).

I'll have to refactor clusterd to use elektra for its database, then I can simply set the keys as I would with a normal instance and push the changes to the instances assigned to the cluster.

markus2330 commented on src/tools/web/doc/daemon_structure.gliffy:
```diff
@@ -0,0 +1 @@
+{"contentType":"application/gliffy+json","version":"1.1","metadata":{"title":"untitled","revision":0,"exportBorder":false},"embeddedResources":{"index":0,"resources":[]},"stage":{"objects":[{"x":441,"y":209,"rotation":0,"id":51,"uid":"com.gliffy.shape.basic.basic_v1.default.line","width":100,"height":100,"lockAspectRatio":false,"lockShape":false,"order":51,"graphic":{"type":"Line","Line":{"stroke ...
```
newlines?

omnidan replied:
there's a program called [gliffy](https://www.gliffy.com/) that generates these diagrams and its storage format seems to be json. I could add newlines, but every time this file is edited with gliffy, it will remove them again, so it's probably best to leave it like that.

markus2330 replied:
Okay, then we can keep it that way.

markus2330 commented on src/tools/web/clusterd/package.json:
```diff
@@ -0,0 +1,45 @@
+  "homepage": "https://github.com/omnidan/elektra-web#readme",
+  "dependencies": {
+    "body-parser": "^1.15.2",
+    "cors": "^2.8.1",
```
please avoid cross-origin resource sharing. All data should be delivered from same node.

omnidan replied:
in that case clusterd would also have to serve the client - and then I still need to figure out a way to do that while keeping the projects in separate folders (it would get a little messy if everything was in the clusterd folder), but it should be possible

omnidan commented:
I fixed some of the issues and [added TODOs](https://github.com/ElektraInitiative/libelektra/projects/4) for the others.

Right now my main problem is setting configuration of a cluster (because I don't use elektra as clusterd's storage). After refactoring clusterd to use elektra, it should be trivial though. Everything else is implemented already.

markus2330 commented:
What is a useful stage to review and accept this PR? Did you finish the refactoring you talked about yesterday?

omnidan commented:
@markus2330 I finished the refactoring of the backend, but there are still some improvements missing: https://github.com/ElektraInitiative/libelektra/projects/4

I would merge the PR when I'm finished with all of these and the implementation is fully consistent with the API documentation. Until then it can stay open for reviews.

Namoshek commented on src/tools/web/clusterd/src/db/utils.js:
```diff
@@ -0,0 +1,31 @@
+  ROOT_PATH + '/' + path
+
+export const virtualKdb = (clusterId, kdbPath) =>
+  path(`vkdb/${clusterId}${kdbPath ? '/' + kdbPath : ''}`)
```
Make path to virtual KDB configurable

Namoshek commented on src/tools/web/clusterd/src/routes/utils.js:
```diff
@@ -0,0 +1,8 @@
+  JSON.stringify(obj, null, 2)
+
+export const successResponse = (res, output) =>
+  output ? res.send(prettyprint(output)) : res.status(404).send()
```
404 in a success response??

Namoshek commented on src/tools/web/clusterd/src/routes/utils.js:
```diff
@@ -0,0 +1,8 @@
+  output ? res.send(prettyprint(output)) : res.status(404).send()
+
+export const errorResponse = (res, err) =>
+  res.status(400).send(prettyprint({ status: 400, message: err.message }))
```
Remove `status` from message body

Namoshek commented on src/tools/web/elektrad/package.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "elektrad",
+  "version": "1.0.0",
+  "description": "",
```
Description

Namoshek commented on src/tools/web/elektrad/src/app.js:
```diff
@@ -0,0 +1,15 @@
+
+import initRoutes from './routes'
+
+const PORT = 1234
```
Make port configurable

Namoshek replied:
This also seems to be still an issue (especially because 1234 is not really an uncommon port).

markus2330 commented on src/tools/web/README.md:
```diff
@@ -0,0 +1,45 @@
+The Web UI allows the user to add new instances to the network, as well as
+combine multiple instances into a cluster. If the configuration of a cluster is
+edited, the changes are pushed to all instances in the cluster. Furthermore,
+single instances can be configured independently.
```
Formatted text (e.g. with tool `fmt` or automatic linebreaks from an editor) have the advantage that diffs on them are much nicer.

krit0n commented on src/tools/web/kdb.js:
```diff
@@ -0,0 +1,116 @@
+const escapeValues = (template, ...values) =>
+  template.reduce((acc, part, i) => {
+    let val = values[i - 1]
+    if (typeof val === 'string') val = `"${val}"`
```
I'm no security expert but I think it is problematic, that strings that get passed to clusterd and elektrad make it to the commandline.
e.g.: I can do that `http GET 'localhost:1234/kdb/user/$(touch $HOME/somefile)'`, which creates a file in my home directory.

At least the values should be escaped with single instead of double quotes here.

krit0n replied:
I can still do this and create a file in my homedir. Is this not a problem?

krit0n commented on src/tools/web/kdb.js:
```diff
@@ -0,0 +1,116 @@
+// import javascript object at given `path`
+const _import = (path, value) =>
+  safeExec(
+    `echo "${JSON.stringify(value).replace(/\"/g, '\\"')}" | ` + // pipe json into kdb
```
This just adds one backslash to a _double quote_. When value is `\"` this _unescapes_ the double quote.
Maybe [this](https://github.com/ElektraInitiative/libelektra/pull/983/files/a45256ee246009e2199590e7642085ec611f6918#r83965059) regular expression can work here.

markus2330 replied:
But hopefully we soon have a NodeJS binding, then we can get rid of these hacks.

krit0n replied:
This regexp still unescapes double quotes. [This](https://github.com/ElektraInitiative/libelektra/pull/983/files/a45256ee246009e2199590e7642085ec611f6918#r83965059) regexp should work here.

omnidan commented on src/tools/web/kdb.js:
```diff
@@ -0,0 +1,116 @@
+  safeExec(escapeValues`kdb rm ${path}`)
+
+// export javascript object from given `path`
+const _export = (path) =>
```
I tried starting `clusterd` with node 6.8.1 and the latest source and it worked fine - maybe it got fixed during the refactoring, can you try again and let me know what happens? If it still fails, please let me know how to reproduce the issue.

krit0n replied:
I tried it - now with node 6.9.2 - and everything worked fine.

krit0n replied:
It still doesn't work.

Here is a Vagrantfile that matches the environment in which i tested your pullrequest (HASH is your pullrequests hash)
```ruby
Vagrant.configure("2") do |config|
  config.vm.box = "ubuntu/xenial64"

  config.vm.provision "shell", inline: <<-SHELL
    apt-get update
    apt-get install -qq -y curl build-essential cmake pkg-config libyajl-dev

    curl -sL https://deb.nodesource.com/setup_6.x | bash -
    apt-get install -qq -y nodejs

    HASH='0823edac81b0a33a53ab321a7cdf1aa456dd80ed'
    curl https://codeload.github.com/ElektraInitiative/libelektra/tar.gz/${HASH} | tar xz --transform s/libelektra-${HASH}/libelektra/
    mkdir libelektra/build && cd $_ && cmake .. && make && make install
    echo "/usr/local/lib" > /etc/ld.so.conf.d/usrlocal.conf
    ldconfig
  SHELL
end
```
If you start nodejs and execute
```node
child_process.exec('kdb export "user/sw/elektra/web/#0/current/clusters" yajl', (err, stdout, stderr) => console.log(stderr)); '';
```
this error is thrown. Although
`kdb export "user/sw/elektra/web/#0/current/clusters" yajl` works in the commandline.

I found this error, because i can't open clusters because of this error.

Namoshek replied:
Could be an issue with kdb not being given by its full path, e.g. `/usr/local/bin/kdb`?

markus2330 replied:
When `kdb` works the PATH should be okay. Unless npm modifies it?

Which error is thrown?

krit0n replied:
The error from my first comment.

omnidan replied:
@krit0n I tried running your Vagrantfile, but it's asking for a password, which I don't know (I tried the common default passwords, including `vagrant`). Might be related to [this issue](https://bugs.launchpad.net/cloud-images/+bug/1569237). Any idea how I can get it running?

krit0n replied:
The xenial64 box  was updated on 21.Dec. I used the previous version where everything worked fine.
You could edit the provided Vagrantfile to use the previous version by adding `config.vm.box_version`:
```ruby
Vagrant.configure("2") do |config|
  config.vm.box = "ubuntu/xenial64"
  config.vm.box_version = "20161130.0.0"
  // ..
end
```
Alternatively you could add the user and password with the config options `config.ssh.username = "ubuntu"` and `config.ssh.password = "bce13c362bdf6d1518510c23"` (I found the password in `~/.vagrant.d/boxes/ubuntu-VAGRANTSLASH-xenial64/20161221.0.0/virtualbox/Vagrantfile`)

I'm sorry for the inconvenience, but I thought providing a vagrant environment is the easiest possibility to reproduce the error.


I hope it works now. If it still doesn't work you might try it with docker:

Place this in a `Dockerfile`:
```bash
FROM ubuntu:xenial

RUN apt-get -qq update && apt-get -qq -y install \
    curl build-essential cmake pkg-config libyajl-dev apt-transport-https

RUN curl -sL https://deb.nodesource.com/setup_6.x | bash - && apt-get install -qq -y nodejs

ARG HASH=0823edac81b0a33a53ab321a7cdf1aa456dd80ed
RUN cd && curl https://codeload.github.com/ElektraInitiative/libelektra/tar.gz/${HASH} | \
    tar xz --transform s/libelektra-${HASH}/libelektra/ && \
    mkdir libelektra/build && cd libelektra/build && cmake .. && make && make install && \
    echo "usr/local/lib" > /etc/ld.so.conf.d/userlocal.conf && ldconfig
```
You can build and run it with:
```bash
docker build --tag nodebug:latest .
docker run -it nodebug bash
```

omnidan replied:
@krit0n thanks a lot for the extensive guide and providing an environment to reproduce the error :ok_hand:

I tried it with the Dockerfile as I'm more familiar with docker than with vagrant, and it worked fine!

Now I could reproduce your issue, it does seem to be related to nodejs, because this works in your environment:

```
> kdb export "user/sw/elektra/web/#0/current/clusters" yajl
{

}
```

While these don't work (via `node` v6.9.4):

```
child_process.exec('kdb export "user/sw/elektra/web/#0/current/clusters" yajl', (err, stdout, stderr) => console.log(stderr)); '';
```

```
child_process.exec('/usr/local/bin/kdb export "user/sw/elektra/web/#0/current/clusters" yajl', (err, stdout, stderr) => console.log(stderr)); '';
```

Both of these commands result in:

```
Error (#75) occurred!
Description: Could not open file for writing.
Ingroup: plugin
Module: storage
At: /root/libelektra/src/plugins/yajl/yajl_gen.c:255
Reason: No such device or address
Mountpoint: user/sw/elektra/web/#0/current/clusters
Configfile: /dev/stdout
```

@markus2330 any idea what the issue could be here?

omnidan commented:
@Namoshek @kirt0n I implemented all your comments, please re-review and let me know if everything is fine now :)

omnidan commented:
@markus2330 I think this can be merged now (let me know before you merge, so I can change the URLs). Still missing is `npm run package` and the cmake integration, I would make a separate PR with these and other improvements (second version of elektra-web).

omnidan commented:
I also cloned the project freshly and tested everything with the latest node version (`7.2.1`).

It's also easier to set up now, because clusterd serves the client.

Terminal 1:

```
cd elektrad
npm install
npm start
```

Terminal 2:

```
cd clusterd
npm install
npm start
```

Web browser: visit http://localhost:1235 and add instance with host `http://localhost:1234`

markus2330 commented:
cmake integration: make sure all files get installed (in tools/data folders) and then run `npm` during installation.

krit0n commented on src/tools/web/kdb.js:
```diff
@@ -0,0 +1,128 @@
+    (source: @krit0n - https://github.com/ElektraInitiative/libelektra/pull/983#discussion_r83965059)
+    */
+    let val = values[i - 1].replace(/((\\\\)*)(\\(")|("))/g, '$1\\$4$5')
+    if (typeof val === 'string') val = `"${val}"`
```
[This](https://github.com/ElektraInitiative/libelektra/pull/983#discussion_r83977227) is still an issue here.

omnidan commented:
@Namoshek I made elektrad use port `33333` by default (not used according to [a list of port numbers](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers)), and clusterd uses `33334` by default now. All ports can be configured via the `PORT` environment variable now (even in `elektrad`).

omnidan commented:
@krit0n I used single quotes `'` and `escapeValues` everywhere (even for `_import`, where previously your regex wasn't being used). Now I am not able to create a file in my homedir via `http GET 'localhost:1234/kdb/user/$(touch $HOME/somefile)'` anymore (I tested it and it worked before the changes).

markus2330 commented:
> `http GET 'localhost:1234/kdb/user/$(touch $HOME/somefile)'`

That sounds like code injection and should be prohibited. Why do you want to do something like that?

omnidan commented:
@markus2330 I tried this with the latest version and it doesn't work anymore now. The problem was that I passed the string via double quotes to the shell (which allows code execution like that). Proper node bindings should avoid such issues in the future.

markus2330 commented:
> Proper node bindings should avoid such issues in the future.

Yes, then lets keep this for later.

@Namoshek @krit0n Are all your points now resolved? Please Approve changes if they are.

markus2330 commented on src/tools/web/client/LICENSE:
```diff
@@ -0,0 +1,21 @@
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
```
Even though MIT actually is similar permissive as BSD, you should explicitely note somewhere that the code is dual licensed (or BSD-only, then simply remove this file). Otherwise users of Elektra would need to copy multiple of license files.

markus2330 commented on src/tools/web/client/src/components/InstanceCard.jsx:
```diff
@@ -0,0 +1,64 @@
+      : name
+
+    return (
+        <Card style={{ margin: '10px', marginBottom: '25px' }}>
```
And is this particular piece actually reusable?

Namoshek commented:
Sorry, I'm not really used to the github review system yet. But it should be fine now, my stuff is resolved!

markus2330 commented on src/tools/web/client/src/components/TreeView.jsx:
```diff
@@ -0,0 +1,72 @@
+// adapted from https://github.com/chenglou/react-treeview
```
No other comments in the file. What did you change? Why not use the treeview? Is yours different enough to claim authorship? (Otherwise we have license troubles)

omnidan replied:
@markus2330 at the moment, I added input fields to tree elements with values (this would not be possible with the original library). As this will be the focus in further iterations of elektra-web, the tree view component will still be expanded with more functionality later. I also added comments about this in the file.

markus2330 commented on src/tools/web/client/src/containers/ConnectedApp.js:
```diff
@@ -0,0 +1,13 @@
+import { connect } from 'react-redux'
```
Every file should at least have some docu.

markus2330 commented on src/tools/web/clusterd/LICENSE:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
See above.

markus2330 commented on src/tools/web/clusterd/README.md:
```diff
@@ -0,0 +1,32 @@
+# clusterd
```
Please add a proper README.md explaining what it does, how it does it and so on..

markus2330 commented on src/tools/web/clusterd/src/config.js:
```diff
@@ -0,0 +1,2 @@
+export const PORT = process.env.PORT || 33334
```
Please fetch settings from kdb, not from env.

omnidan replied:
@markus2330 it is a sort-of standard to allow applications to be executed on a different port by specifying the `PORT` envvar (e.g. a lot of container systems work with this). does it really make sense to load it from elektra? maybe try to load it from the env var first, then elektra, then fall back to the default port?

markus2330 commented on src/tools/web/clusterd/src/connector.js:
```diff
@@ -0,0 +1,39 @@
+import fetch from 'node-fetch'
```
No comments..

markus2330 commented on src/tools/web/elektrad/LICENSE:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
again..

markus2330 commented on src/tools/web/elektrad/README.md:
```diff
@@ -0,0 +1,31 @@
+# elektrad
```
again..

markus2330 commented:
@Namoshek thanks for the quick response. I wondered why your virtual kdb comment still was in the "File Changed" section. Good that everything is resolved now, thanks for your review!

krit0n commented:
The issues I found in the last review are resolved now. (except the one [here](https://github.com/ElektraInitiative/libelektra/pull/983#discussion_r83937165), but i seem to be the only one that is affected by this)

markus2330 commented:
> Reason: No such device or address
> Configfile: /dev/stdout
> @markus2330 any idea what the issue could be here?

Seems like /dev/stdout does not exist within nodejs. There is an undocumented 
feature that you can pass a different file name as 3rd command-line option to 
`kdb export`. You should not rely on this feature but as workaround until 
proper NodeJS bindings are written it should suffice.

markus2330 commented:
@omnidan This is still open, @krit0n cannot check this as long the `/dev` issue is not fixed.

omnidan commented:
@markus2330 @krit0n sorry it took so long! the issue is fixed now (by writing to a buffer file and reading that instead of using stdout), I also did a fresh install on macOS and with your docker container to make sure everything works fine. I've improved error handling for certain cases (e.g. when `kdb` is installed but without the `yajl` plugin). I also made everything production ready, all devtools get disabled in production now.

I've also started working on refactoring the react components to make them more reusable (in the [elektra-web-update branch](https://github.com/omnidan/libelektra/compare/1e1f389...elektra-web-update)) and I'll work on adding more documentation.

BTW: there is an issue with the `postinstall` hook when running everything as root (like in @krit0n's docker container) - because [npm tries to downgrade privs](https://github.com/npm/npm/issues/3497), it will fail to install the client dependencies (which are in a separate directory than the working directory). a workaround for this is simply not running `npm` as root, or running `npm install` manually in the `client/` directory.

markus2330 commented:
> @markus2330 it is a sort-of standard to allow applications to be executed on
> a different port by specifying the `PORT` envvar (e.g. a lot of container
> systems work with this). does it really make sense to load it from elektra?
> maybe try to load it from the env var first, then elektra, then fall back
> to the default port?

Elektra will also allow you to use envvars for its configuration (see #1252) 
so it would influence the lookup in unexpected ways if you also add support 
for it.

But you can leave it now as-is and provide support later when you properly 
elektrify your application.

omnidan commented:
@markus2330 alright, then I will load the configuration from elektra once I have proper node.js bindings.

I've also added more documentation (README.md's and in the source code). The PR should be ready to merge now. I've already replaced all links except for the images in the main README.md:

```
https://cdn.rawgit.com/omnidan/libelektra/elektra-web/src/tools/web/doc/
```

These links need to be replaced with:

```
https://cdn.rawgit.com/ElektraInitiative/libelektra/master/src/tools/web/doc/
```

(Which will be broken until the PR is merged)

markus2330 commented on src/tools/web/clusterd/src/app.js:
```diff
@@ -1,3 +1,7 @@
+/* app.js
```
Please do not repeat the file name as content within the file.

Please use Elektra's standard header, see `doc/CODING.md`.

omnidan replied:
@markus2330 done!

markus2330 commented:
Thank you!

I was playing around a bit and created  #1314.
Looking forward to further improvements!
